pandas/core/internals/construction.py

```diff
@@ -351,14 +351,19 @@ def ndarray_to_mgr(
             # transpose and separate blocks
 
             dtlike_vals = [maybe_infer_to_datetimelike(row) for row in values]
-            dvals_list = [ensure_block_shape(dval, 2) for dval in dtlike_vals]
-
-            # TODO: What about re-joining object columns?
-            block_values = [
-                new_block(dvals_list[n], placement=n, ndim=2)
-                for n in range(len(dvals_list))
-            ]
-
+            # don't convert (and copy) the objects if no type inference occurs
+            if any(
+                not is_dtype_equal(instance.dtype, values.dtype)
+                for instance in dtlike_vals
+            ):
+                dvals_list = [ensure_block_shape(dval, 2) for dval in dtlike_vals]
+                block_values = [
+                    new_block(dvals_list[n], placement=n, ndim=2)
+                    for n in range(len(dvals_list))
+                ]
+            else:
+                nb = new_block(values, placement=slice(len(columns)), ndim=2)
+                block_values = [nb]
```

Comment:
can we remove the entire if/then here e.g. L350 and just use the inside logic (inlucding what you added), e.g. as L368 look duplicative now.